File: Medium/1143. Longest Common Subsequence/solution (4).py
# 1143. Longest Common Subsequence

# Runtime: 2496 ms, faster than 5.01% of Python3 online submissions for Longest Common Subsequence.

# Memory Usage: 24.7 MB, less than 25.08% of Python3 online submissions for Longest Common Subsequence.


class Solution:
    # Dynamic Programming
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        lens = [[0] * (len(text2) + 1) for _ in range(len(text1) + 1)]

        def lcs(i: int, j: int) -> int:
            if i == 0 or j == 0:
                return 0
            elif lens[i][j] > 0:
                return lens[i][j]

            if text1[i - 1] == text2[j - 1]:
                lens[i][j] = 1 + lcs(i - 1, j - 1)
            else:
                lens[i][j] = max(lcs(i, j - 1), lcs(i - 1, j))
            return lens[i][j]

        return lcs(len(text1), len(text2))

File: Medium/740. Delete and Earn/solution (4).cpp
// 740. Delete and Earn

// Runtime: 51 ms, faster than 12.83% of C++ online submissions for Delete and Earn.

// Memory Usage: 13.1 MB, less than 60.14% of C++ online submissions for Delete and Earn.


class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        initPoints(nums);
        vector<int> maxPoints(uniqueNums_.size() + 1);
        maxPoints[1] = getPoint(uniqueNums_[0]);

        for (int i = 2; i < maxPoints.size(); i++) {
            if (uniqueNums_[i - 1] != uniqueNums_[i - 2] + 1) {
                maxPoints[i] = maxPoints[i - 1] + getPoint(uniqueNums_[i - 1]);
            } else {
                maxPoints[i] =
                    max(maxPoints[i - 1],
                        maxPoints[i - 2] + getPoint(uniqueNums_[i - 1]));
            }
        }

        return maxPoints.back();
    }

private:
    void initPoints(vector<int>& nums) {
        for (int n : nums) {
            points_[n] += n;
        }

        for (auto [n, _] : points_) {
            uniqueNums_.push_back(n);
        }
    }

    int getPoint(int num) const {
        if (auto point {points_.find(num)}; point != points_.end()) {
            return point->second;
        } else {
            return 0;
        }
    }

    map<int, int> points_;
    vector<int> uniqueNums_;
};